### ISSUE TITLE
Some questions about EVA pretraining

### ISSUE THREAD

**Issue by conicoco1993** [baaivision/EVA#28]
1. In your opinion, is EVA a method of both model scaling and data scaling? Does pretraining with more data (such as the data used in CLIP finetuning) yield better results than using only the 30M data described in the paper? What data is optimal for EVA?
2. How does the teacher model influence the performance of the student model? Is it possible to replace the CLIP model with a supervised model trained on IN21k?
3. In #19, you mentioned that ViT has many desirable properties, but some researchers have observed that scaling the ViT model can lead to instability during fp16 training (such as overflow in the forward pass and underflow in the backward pass). Why do you continue to choose ViT as a backbone over alternatives like SwinTransformer, and could you elaborate on ViT's "good properties"? Thank you.

**Comment by Yuxin-CV** [comment]
Hi @conicoco1993, thanks for your interest in EVA and your valuable questions.

> Q1: In your opinion, is EVA a method of both model scaling and data scaling? Does pretraining with more data (such as the data used in CLIP finetuning) yield better results than using only the 30M data described in the paper? What data is optimal for EVA?

A1: We didn't conduct a detailed / direct study on data scaling using the 1.0B EVA, but we have some observations that it is not always better to use more pre-training data. Noisy data would cause optimization issue during very large model pre-training. E.g., we find when using a 10M subset of LAION-400M image data for EVA MIM pre-training, the pre-training loss is relatively higher compared with using the 14M IN-21K, and has more "loss spikes". The Google's T5 also conducts a study on the pre-training dataset size & quality (on a based-sized model only) and yields similar conclusion: large noisy data could be harmful, and there exists a (secret) "sweet point" (Meanwhile, CLIP is different, CLIP can be scale up w/ large-scale noisy image-text paired data, so CLIP is great.)

From another perspective, we study EVA-L (304M parameters) using IN-21K as pre-training data as well as EVA-CLIP features as MIM targets (https://github.com/baaivision/EVA/tree/master/eva#eva-l-learning-better-mim-representations-from-eva-clip), the resulting ViT-L can outperform the previous art MIM model (e.g, BEiTv2-ViT-L) by a large margin: directly fine-tune MIM pre-trained (w/o supervised intermediate fine-tuning, which is a hack for some specific benchmarks and interfere the study of MIM pre-training) EVA-L achieves 88.0 acc1 @ 196px & psz14 (+0.7 compared w/ the BEiTv2 counterpart using a similar hyper-params set). And now we have some updated results that the same MIM only pre-trained EVA-L can achieve 88.4 acc1 @ 196px & psz14 (+1.1 compared w/ the BEiTv2!) with better tuned fine-tuning hyper-params. These results indicate that w/ better target (EVA-CLIP) and more data (IN-21K) **together**, the performance can be largely improved. We are now running jobs w/ EVA-CLIP & IN-1K for a clearer study of pre-training data (IN-21K v.s. IN-1K) on the larged-sized model.



> Q2: How does the teacher model influence the performance of the student model? Is it possible to replace the CLIP model with a supervised model trained on IN-21K?

A2: We didn't conduct a detailed / direct study on that, and unfortunately we cannot answer this question directly. But based on our study, we believe the scalability of MIM pre-training is closely related to the scalability of the teacher / MIM prediction target. If the teacher is more scalable, the MIM pre-training is also more scalable. But we don't have a directly study / head-to-head comparison on that. 

BTW, we believe using the word "teacher" is inaccurate to some extent, because when you use the word "teacher", you are talking about the pre-training from a knowledge distillation perspective. But knowledge distillation itself is NOT scalable: you can hardly train a student even stronger than the teacher, while EVA (1.0B) can outperform its teacher (CLIP-L, 304M) by a large margin. On the other hand, MIM prediction on pixel-space is also NOT (data) scalable (https://arxiv.org/abs/2206.04664). However, MIM w/ CLIP feature is scalable. This is really interesting and deserve further study.



> Q3: In #19, you mentioned that ViT has many desirable properties, but some researchers have observed that scaling the ViT model can lead to instability during fp16 training (such as overflow in the forward pass and underflow in the backward pass). Why do you continue to choose ViT as a backbone over alternatives like SwinTransformer, and could you elaborate on ViT's "good properties"?

A3: To my knowledge and from my option, vanilla ViT is the MOST scalable architecture to date. Compared to ANY other architectures, vanilla ViT is always easier to (pre-)train. Because 
1. Vanilla ViT has fewer structure prior, and learns nearly everything from the pretext task. While structure prior in other architectures may incompatible w/ the pretext task. On the other hand, task prior can be also learned from pretext task. 
2. Vanilla ViT can be easily pre-trained w/ MIM, and MIM is (conditionally) quite scalable. 
3. The NLP community has already shown that, even w/ fp16, vanilla ViT can scalable up to over 100B parameters (e.g., [GLM-130B](https://github.com/THUDM/GLM-130B), [OPT](https://arxiv.org/abs/2205.01068)), and consistent performance gain is observed during model scaling. So if you encounter instability issue, it is not the fault of vanilla ViT, and please always believe that our CVer can solve it. 

So if I can only pre-train a very large model once, I choose vanilla ViT, and If I can pre-train a very large model twice / thrice..., I still choose vanilla ViT.

**Comment by conicoco1993** [comment]
Thank you very much for the author's response, I now have a deeper understanding of EVA. I still have some questions that I would like to discuss with you:

Through your answer to question 1, I understand that clean data is very important for EVA.
a. Does this mean that EVA's data scalability only works with clean data and is not effective for noisy data?
b. Is there any way to filter out more clean data from datasets like LAION, in order to further improve the effectiveness of EVA's pre-training?
c. I am very confused about how noisy is defined for unlabeled data?
Through the answer to question 2, I understand that MIM+CLIP feature has some "magical" effects (even though the CLIP feature is obtained from a smaller model). The method of modeling the relevance of high-level features through context prediction is already very similar to pre-training in the NLP field. However, I feel that the difference is that these high-level features in EVA are "expensive" and "infinite (continuous)".
a. I am quite curious about how much EVA is currently constrained by the effects of the EVA-CLIP model, and also in the future?
b. Do you have any insights on ways to obtain "free" high-level features in the visual field?

**Comment by Yuxin-CV** [comment]
Thanks for your valuable questions. 

> Q1 (c): I am very confused about how noisy is defined for unlabeled data?

A: This is a very good question: what makes for good data for MIM pre-training? In the context of my previous response, "noisy" data refers to: under similar dataset size (i.e., ~11M images randomly sampled from LAION-400M), the MIM training loss is higher & more unstable compared with using IN-21K data (~14M images). This is definitely not a good definition for "noisy" data. But from another perspective, IN-21K is more or less an organized and curated dateset, while LAION-400M is random and un-curated. Study the scale / distribution / quality of pre-training data is indeed meaningful and is lack in current research, but this is beyond the scope of EVA and require much more budgets.

For reference, [SEER](https://github.com/facebookresearch/vissl/blob/main/projects/SEER/README.md) studies this topic, but I don't think the scaling property / performance is very satisfactory, at least among the tasks of my interest. So in EVA, we merge some well-known & more or less curated datasets for pre-training, e.g., IN-21K, CC-12M, CC-3M, Object365, etc. We don't konw whether this is the best choice, but the performance looks good. 

> Q1 (a): Does this mean that EVA's data scalability only works with clean data and is not effective for noisy data?

A: We don't think the MIM pre-training data is very "clean", because we directly merge the available datasets for pre-training, and we don't tune the sampling ratio of different datasets, etc. This could be potentially harmful, because the data distribution is a directly merged distribution, which would not be so "natural".

We plan to conduct a study of the MIM pre-training data (rangng from 0.1M to tens of millions) on the large-sized model in the future.

> Q1 (b): Is there any way to filter out more clean data from datasets like LAION, in order to further improve the effectiveness of EVA's pre-training?

A: This is a very good question and we don't know and definitely deserves an independent research work. To my knowledge, the image-text pair is filtered with OpenAI‘s [CLIP](https://openai.com/blog/clip/) by calculating the cosine similarity between the text and image embeddings and dropping those with a similarity below 0.3. But still this large-scale dataset is non-curated, especially from the perspective of unlabeled images only.

**Comment by Yuxin-CV** [comment]
> The method of modeling the relevance of high-level features through context prediction is already very similar to pre-training in the NLP field. However, I feel that the difference is that these high-level features in EVA are "expensive" and "infinite (continuous)".

Response: I don't think the feature from CLIP or EVA-CLIP is "expensive". On the contrary, the CLIP / EVA-CLIP feature is (nearly) "free". Because CLIP is a real foundation model and can be used in many other tasks across different modalities in a zero-shot manner. So once you have a really good CLIP, you can use it infinite times without cost / tedious fine-tuning, and you only need to train a CLIP once. Leveraging CLIP features for MIM pre-training is only one usage among many other tasks that CLIP capable of, so that why we choose to train a giant CLIP and open-source it.

As for "infinite (continuous)", I think this is actually an advantage compared with the discrete tokenization in NLP. The tokenizer constraints the MLM input & output space to be fixed & discrete, which could cause a representation bottleneck. While pixel and visual feature is more flexible and general.

This could also explain why EVA / EVA-L can outperform BEiT3 / BEiT2 in some tasks to some extent: BEiT2 / BEiT3 leverages tokenizer for MIM prediction targets, the tokenizer produces discrete outputs, and is trained on IN-1K data only therefore has IN-1K biases which could be harmful for generalization. Another issue of using tokenizer is that it is not scalable. Meanwhile CLIP is scalable, and could potentially make MIM pre-training scalable.

Please also refer to some recent research work / discussion in NLP on this topic: 
- Language Modelling with Pixels ([paper](https://arxiv.org/abs/2207.06991))
-  Image-and-Language Understanding from Pixels Only ([paper](https://arxiv.org/abs/2212.08045), [discussion](https://twitter.com/chriswolfvision/status/1604402429042524160))

**Comment by Yuxin-CV** [comment]
> Q2 (a): I am quite curious about how much EVA is currently constrained by the effects of the EVA-CLIP model, and also in the future?

A: To clarify, EVA (~1.0B) is pre-trained with MIM using OpenAI CLIP-L vision tower (~304M), and EVA-CLIP is a 1.1B CLIP trained by us with MIM pre-trained EVA as the initialization of the vision tower, which can be also served as MIM pre-training targets.

We believe by scaling CLIP, the performance of MIM pre-training with CLIP feature targets will also be consistently improved. So if MIM pre-training is "constrained" by CLIP, and can be improved by using a better CLIP, then it is a good news for MIM. Because better CLIP is easy to obtain: you can simply improve CLIP by scaling up it. 

> Q2 (b): Do you have any insights on ways to obtain "free" high-level features in the visual field?

To date, open-sourced CLIPs ([OpenAI CLIP](https://github.com/openai/CLIP), [Open CLIP](https://github.com/mlfoundations/open_clip), [EVA-CLIP](https://github.com/baaivision/EVA/tree/master/clip)) are the best "free" high-level features in the visual field.

**Comment by conicoco1993** [comment]
Thank you for your answer. Feel free to close the issue

**Comment by Yuxin-CV** [comment]
Thanks for your interest in EVA and your valuable questions. I'm closing this issue. Feel free to ask if you have further questions.

**Comment by LiJunnan1992** [comment]
Congrats on the great work! EVA-CLIP is performing really well for our BLIP-2 models, thanks for the open-source!

Your above explanations help clear a lot of my doubts. I have a few follow-up questions:

1. Regarding the definition of "noise" for MIM. In LAION, there exist many images with white background. This might make the MIM target to be uninformative and lead to instability in optimization. This is just my unverified hypothesis.
2. Do you have any plans in further improving the EVA models? Maybe using EVA-CLIP as the "teacher" to train a ViT-G model?

Thanks,
Junnan

**Comment by Yuxin-CV** [comment]
Hi @LiJunnan1992, thanks for using EVA in the awesome BLIP-2. Regarding to your questions:

1. We agree with your opinion on using LAION in MIM. So currently, we are still using merged, curated datasets in MIM pre-training. But we also believe LAION is a treasure for MIM, as long as filtered rightly.

2. We are building a surprisingly strong visual representation, and it will be released soon, please stay tuned.

**Comment by LiJunnan1992** [comment]
Looking forward to your new model! Can I know an estimated timeline of the release?

**Comment by Yuxin-CV** [comment]
> Looking forward to your new model! Can I know an estimated timeline of the release?

Before the end of March, we will release the full suite of the next gen EVA.

**Comment by Yuxin-CV** [comment]
Hi @LiJunnan1992, if you are still interested in better CLIP models powered by EVA-01 / -02, please refer to https://github.com/baaivision/EVA/tree/master/EVA-CLIP#model-card. 

We offer performant CLIP models at all scales, range from 150M to 4.7B. The model weights as well as the inference code were just released.

**Comment by LiJunnan1992** [comment]
Thanks @Yuxin-CV ! These new models look amazing. I'll check them out.
